Factor out TensorBase that doesn't depend on native operators

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64499
* #64173
* #64172
* #64171
* #64170
* #64169
* #64168
* #64167
* #64166
* #64165
* __->__ #63612

This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

Differential Revision: [D30728580](https://our.internmc.facebook.com/intern/diff/D30728580)

peterbell10 committed:
Factor out TensorBase that doesn't depend on native operators

This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/NamedTensor.cpp:
```diff
@@ -1,6 +1,7 @@
+#define TORCH_ASSERT_NO_OPERATORS
```
We should do this more often; I'm super annoyed NativeFunctions.h is accessible from ATen.h

peterbell10 replied:
Oh interesting, that really shouldn't be public. At this point I guess it would break the world to remove though?

ezyang replied:
It will be somewhat disruptive, as people often do `native::blah` to bypass the dispatcher, but I think it's worth fixing

ezyang commented:
@zdevito mentions that one problem with having TensorBase and Tensor is that if you have `std::vector<Tensor>` you have to reallocate a new tensor `std::vector<TensorBase>` to appease the typechecker (since template parameters in C++ are invariant). Did you have any thoughts about this?

ezyang commented:
One happy side effect of this PR is that TensorBase being factored out into its own class will make it easier to write other classes like CPUTensor/CUDATensor

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728580).

ezyang commented on aten/src/ATen/core/VariableHooksInterface.h:
```diff
@@ -40,23 +40,25 @@ namespace impl {
+  virtual int64_t _version(const TensorBase&) const = 0;
+  virtual void retain_grad(const TensorBase&) const = 0;
+  virtual bool retains_grad(const TensorBase&) const = 0;
   virtual void _backward(const Tensor&, TensorList, const c10::optional<Tensor>&, c10::optional<bool>, bool) const = 0;
```
This is an example where invariance of containers means you can't replace Tensor with TensorBase. Is there a downside to not converting the signature here?

peterbell10 replied:
I might look into it if it becomes relevant, but at the moment I'm focused on operator kernels. Kernels both take the largest proportion of the compile time, and also don't really use autograd so it isn't relevant here.

The above comment about needing to reallocate to convert between `std::vector<Tensor>` and `std::vector<TensorBase>` is also an important factor here.

ezyang replied:
That's fine. I was just a little surprised that it was OK not to update this haha.

peterbell10 replied:
This works out because `VariableHooksInterface.h` is only included from the `Tensor.cpp` file, not any headers. So, it's okay to include `Tensor.h`.

ezyang commented on aten/src/ATen/templates/Operators.h:
```diff
@@ -2,6 +2,13 @@
+  meaning the file will need to be re-compiled every time an operator      \
+  is changed or added. Consider if your change would be better placed in   \
+  another file, or if a more specific header might achieve the same goal.
+#endif
```
So, @gchanan mentioned that if the problem is just the methods on Tensor, as long as you just don't edit methods in native_functions.yaml (hard, I know) this shouldn't induce recompilation. So then *technically* you don't need to split Tensor and TensorBase; you just need to make sure you aren't accidentally including Functions.h. I appreciate that the situation on the ground, however, might make it a bit easier, but I'm wondering if there is something else flawed about this analysis.

peterbell10 replied:
At the moment, `Tensor.h` does depend on all signatures through `Operators.h` but in theory of course that could be fixed. However, that's still more dependency than is necessary for most operator kernels so either way I think this is valuable.

ezyang reviewed (approved):
I'm willing to give this a shot. Even if the public inheritance is a problem, I'd be willing to keep the split and make the inheritance private, because it makes other things easier. I'm going to let this stew a little more, but if any other reviewers have objections, speak now or forever hold your peace.

peterbell10 commented:
>I'd be willing to keep the split and make the inheritance private, because it makes other things easier.

Can you expand on this? What problems does the inheritance cause?

ezyang commented:
> Can you expand on this? What problems does the inheritance cause?

It is mostly the potential for TensorBase v. Tensor confusion, and people doing conversions to shut up the type checker. I definitely agree that right now, as this PR is right now, there are no problems, I'm just not sure if it will stay this way as people who are less familiar with the architecture here start adding new features. (But I'm willing to give it a try!)

peterbell10 commented:
>It is mostly the potential for TensorBase v. Tensor confusion, and people doing conversions to shut up the type checker.

I had thought of this and it's why I marked the `TensorBase` -> `Tensor` conversion `explicit`, so it's at least visible in code review. However, it's really the `const TensorBase&` arguments that are enabled by inheritance, and those should be painless IMO. Users can call them with a `Tensor` as normal, and the type checker is completely happy.

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

Differential Revision: [D30728580](https://our.internmc.facebook.com/intern/diff/D30728580)

[ghstack-poisoned]

albanD commented on aten/src/ATen/templates/Operators.h:
```diff
@@ -2,6 +2,13 @@
+#error This change adds a dependency on native_functions.yaml,             \
+  meaning the file will need to be re-compiled every time an operator      \
+  is changed or added. Consider if your change would be better placed in   \
+  another file, or if a more specific header might achieve the same goal.
```
I'm not sure that this error message will be helpful for people that don't have context on this PR.
Maybe we want to extend a bit the comment next to TensorBase definition and link it from here.

albanD commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -112,21 +112,23 @@ void AutogradMeta::set_fw_grad(const Variable& new_grad_, const Variable& self,
   } else {
     // TODO(alband) remove this spurious version counter bump
-    auto new_grad = new_grad_;
+    Tensor new_grad(new_grad_base);
```
Is the comment above still valid?

peterbell10 replied:
This PR shouldn't change the result of this code in any way, so afaik it's still relevant.

ezyang replied:
idk, let's ask the guy who added it :P

albanD replied:
The TensorBase -> Tensor still does a version counter bump right?

peterbell10 replied:
It does a reference count bump; not a version count bump.

albanD reviewed (commented):
Thanks for cleaning some autograd namings by side effect :)

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

Differential Revision: [D30728580](https://our.internmc.facebook.com/intern/diff/D30728580)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728580).

ezyang commented:
Some internal builds failing with:

```
buck-out/gen/6ad9e6a1/xplat/caffe2/gen_aten/core/TensorBody.h:5052:36: error: 'unsafe_borrow_t' is a protected member of 'at::TensorBase'
return borrow_type(at::TensorBase::unsafe_borrow_t{}, from); 
                                   ^
buck-out/gen/6ad9e6a1/xplat/caffe2/gen_aten/core/TensorBody.h:5059:36: error: 'unsafe_borrow_t' is a protected member of 'at::TensorBase'
(lhs = borrow_type(at::TensorBase::unsafe_borrow_t{}, rhs)); 
                                   ^
buck-out/gen/6ad9e6a1/xplat/caffe2/gen_aten/core/TensorBody.h:5052:36: error: 'unsafe_borrow_t' is a protected member of 'at::TensorBase'
return borrow_type(at::TensorBase::unsafe_borrow_t{}, from); 
                                   ^
buck-out/gen/6ad9e6a1/xplat/caffe2/gen_aten/core/TensorBody.h:5059:36: error: 'unsafe_borrow_t' is a protected member of 'at::TensorBase'
(lhs = borrow_type(at::TensorBase::unsafe_borrow_t{}, rhs)); 
```

investigating

ezyang committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

Differential Revision: [D30728580](https://our.internmc.facebook.com/intern/diff/D30728580)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728580).

peterbell10 committed:
Update on "Factor out TensorBase that doesn't depend on native operators"


This makes Tensor inherit from a new class TensorBase, that provides a subset of Tensor that doesn't
directly depend on native_functions.yaml. Code that only includes TensorBase.h with thus not need to
be rebuilt every time someone changes an operator signature.

Making `Tensor` inherit from this class means that `const TensorBase&` parameters will be callable
with an ordinary `Tensor`. I've also made `Tensor` constructible and assignable from `TensorBase` to
minimize friction in code mixing the two types.

To help enforce that `Tensor.h` and `Functions.h` aren't accidentally included, I've added an error
into `Operators.h` if `TORCH_ASSERT_NO_OPERATORS` is defined. We can either set this in the build
system for certain folders, or just define it at the top of any file.

I've also included an example of manually special-casing the commonly used `contiguous` operator.
The inline function's slow path defers to `TensorBase::__dispatch_contiguous` which is defined in
`Tensor.cpp`. I've made it so `OptionalTensorRef` is constructible from `TensorBase`, so I can
materialize a `Tensor` for use in dispatch without actually increasing its refcount.

Differential Revision: [D30728580](https://our.internmc.facebook.com/intern/diff/D30728580)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/core/TensorBase.h:
```diff
@@ -83,6 +83,7 @@ class TORCH_API TensorBase {
   explicit TensorBase(unsafe_borrow_t, const TensorBase& rhs)
       : impl_(c10::intrusive_ptr<at::TensorImpl, UndefinedTensorImpl>::reclaim(rhs.impl_.get())) {}
   friend MaybeOwnedTraits<TensorBase>;
+  friend OptionalTensorRef;  // so it can get at unsafe_borrow_t
```
This doesn't work because `OptionalTensorRef` isn't declared yet.

ezyang replied:
RIP. Well, let's just make it public as you did.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728580).
